### DISCUSSION TITLE
CURLOPT_RESOLVE is not working correctly?

### DISCUSSION THREAD

**Discussion by ghost** [curl/curl#8109] [Q&A] [answered]
Hello,

I am trying to use the CURLOPT_RESOLVE option to request a direct IP address.

The documentation says that: https://curl.se/libcurl/c/CURLOPT_RESOLVE.html

> Each single name resolve string should be written using the format [+]HOST:PORT:ADDRESS[,ADDRESS]... where HOST is the name libcurl will try to resolve, PORT is the port number of the service where libcurl wants to connect to the HOST and ADDRESS is one or more numerical IP addresses.

and an example is given:

>    host = curl_slist_append(NULL, "example.com:80:127.0.0.1");
>   curl_easy_setopt(curl, CURLOPT_RESOLVE, host);
>   curl_easy_setopt(curl, CURLOPT_URL, "https://example.com");
 
However - this does not work as described in the documentation.

port number added at - curl_slist_append(NULL, "example.com:80:127.0.0.1") - not working. I can specify the port number at least 666, 
But libcurl, it will connect ONLY by the port number specified in - curl_easy_setopt(curl, CURLOPT_URL, "https://example.com") - that is, the prefix "http" or "https" - only 80 и 443. 

Is this a bug in the documentation or a bug in libcurl?
Or am I doing something wrong?

**Comment by dfandrich** [comment]
On Tue, Dec 07, 2021 at 12:41:53AM -0800, Lator wrote:
> port number added at - curl_slist_append(NULL, "example.com:80:127.0.0.1") -
> not working.

What do you mean by "not working"?

> But libcurl, it will connect ONLY by the port number specified in -
> curl_easy_setopt(curl, CURLOPT_URL, "https://example.com") - that is, the
> prefix "http" or "https" - only 80 и 443.

The resolved address specified above is for port 80, but this URL specifies
port 443 (implicitly). So, the cache won't be used for this request.

**Reply by ghost** [reply]
I can write:

host = curl_slist_append (NULL, «example.com:666:140.82.121.4»);  
curl_easy_setopt (завиток, CURLOPT_RESOLVE, хост);
curl_easy_setopt (curl, CURLOPT_URL, « https://example.com »);

But libcurl - won't connect to port 666. Why then specify this port?

**Comment by jzakrzewski** [comment, marked as answer]
You've understood it backwards.
When you say "example.com:666:140.82.121.4" it means to curl "whenever you want to connect to 'example.com' on port 666, use '140.82.121.4' as the IP". It does not magically change the default port of http/s to 666.
So this entry will only have an effect if you say `curl_easy_setopt (curl, CURLOPT_URL, "https://example.com:666");`

**Comment by jzakrzewski** [comment]
You ONLY need thi if for some reason YOU decide that "example.com:666" should map to "140.82.121.4". Otherwise curl will simply query the DNS for the IP of "example.com". `CURLOPT_RESOLVE` is a user override of that mechanics.
So if you want to connect to different port, you specify it in the URL. If you want to override the name <--> IP mapping, use `CURLOPT_RESOLVE`.

**Reply by ghost** [reply]
> Это нужно вам ТОЛЬКО в том случае, если по какой-то причине ВЫ решите, что "example.com:666" должен отображаться на "140.82.121.4". В противном случае curl просто запросит у DNS IP-адрес example.com. `CURLOPT_RESOLVE`- это переопределение пользователем этой механики. Поэтому, если вы хотите подключиться к другому порту, вы указываете его в URL-адресе. Если вы хотите переопределить отображение имени <--> IP, используйте `CURLOPT_RESOLVE`.

Got it, thanks for the explanation.
